Python/Product/Debugger/DebugAdapterRemoteProcess.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Debugger.DebugAdapterHost.Interfaces;
+using Microsoft.PythonTools.Infrastructure;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.PythonTools.Debugger {
+    sealed class DebugAdapterRemoteProcess : ITargetHostProcess, IDisposable {
+        private const int _debuggerConnectionTimeout = 5000; // 5000 ms
+        private Stream _stream;
+        private bool _debuggerConnected = false;
+
+        private DebugAdapterRemoteProcess() {}
+
+        public static ITargetHostProcess Attach(string attachJson) {
+            var process = new DebugAdapterRemoteProcess();
+            process.AttachProcess(attachJson);
+            return process;
+        }
+
+        private void AttachProcess(string attachJson) {
+            var json = JObject.Parse(attachJson);
+            var uri = new Uri(json["remote"].Value<string>());
+            ConnectSocket(uri);
+        }
+
+        private void ConnectSocket(Uri uri) {
+            var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.IP);
+            EndPoint endpoint;
+            if (uri.IsLoopback) {
+                endpoint = new IPEndPoint(IPAddress.Loopback, uri.Port);
+            } else {
+                endpoint = new DnsEndPoint(uri.Host, uri.Port);
+            }
+            Debug.WriteLine("Connecting to remote debugger at {0}", uri.ToString());
+            var connection = Task.Factory.FromAsync(
+                (AsyncCallback callback, object state) => socket.BeginConnect(endpoint, callback, state),
```

Comment:
Take a look at `SocketTaskExtensions`.

Reply:
That is available in 4.7. Should I switch the framework version?

Reply:
I think it's okay, since VS itself is 4.7. @zooba?

Reply:
I don't know that's the case actually, pretty sure VS still only requires 4.6. Eventually it'll get onto 4.7, but not sure what the timeline is.

Reply:
It is not 4.7 and I'm not sure it will be easy to use 4.7 with VS. I've tried couple months ago to switch RTVS, there were tons of assembly load issues.